File: app/src/main/java/com/ducktapedapps/updoot/UpdootApplication.java
package com.ducktapedapps.updoot;

import android.app.Application;
import android.content.SharedPreferences;

import com.ducktapedapps.updoot.di.ApplicationModule;
import com.ducktapedapps.updoot.di.DaggerUpdootComponent;
import com.ducktapedapps.updoot.di.UpdootComponent;
import com.ducktapedapps.updoot.utils.constants;

import java.util.UUID;

public class UpdootApplication extends Application {
    private UpdootComponent mUpdootComponent;

    @Override
    public void onCreate() {
        super.onCreate();
        getUpdootComponent().inject(this);
        createDeviceId();
    }


    private void createDeviceId() {
        SharedPreferences sharedPreferences = getUpdootComponent().getSharedPreferences();
        if (sharedPreferences.getString(constants.DEVICE_ID_KEY, null) == null) {
            sharedPreferences
                    .edit()
                    .putString(constants.DEVICE_ID_KEY, UUID.randomUUID().toString())
                    .apply();
        }
    }

    public UpdootComponent getUpdootComponent() {
        if (mUpdootComponent == null) {
            mUpdootComponent = DaggerUpdootComponent
                    .builder()
                    .applicationModule(new ApplicationModule(this))
                    .build();
        }
        return mUpdootComponent;
    }
}

File: app/src/main/java/com/ducktapedapps/updoot/UpdootApplication.kt
package com.ducktapedapps.updoot

import android.app.Application
import androidx.work.Configuration
import dagger.hilt.android.HiltAndroidApp
import timber.log.Timber
import javax.inject.Inject

@HiltAndroidApp
class UpdootApplication : Application(), Configuration.Provider {
    @Inject
    lateinit var workConfiguration: Configuration

    override fun getWorkManagerConfiguration(): Configuration = workConfiguration
    override fun onCreate() {
        super.onCreate()
        plantTimberDebugTree()
    }

    private fun plantTimberDebugTree() {
        if (BuildConfig.DEBUG) Timber.plant(Timber.DebugTree())
    }
}